onnxscript/rewriter/ort_fusions/shape_optimization.py

```diff
@@ -17,15 +18,19 @@ def __init__(self):
     It can be simplified away.
     """
 
-    def pattern(self, op, x, dim0, dim1, dim2, dim3):
+    def pattern(self, op, x, dim0, dim1, dim2, dim3, start, end):
         shape = op.Concat(dim0, dim1, dim2, dim3, axis=0)
-        reshaped = op.Reshape(x, shape, allowzero=0)
+        # Note: The allowzero=1 attribute enables us to infer that the shape of the
+        # reshaped tensor is the same as the value of the shape parameter below.
+        # Otherwise, we need to know that there are no zeros in the value of "shape"
+        # for this optimization to be valid.
+        reshaped = op.Reshape(x, shape, allowzero=1)
         transposed = op.Transpose(reshaped, perm=[0, 2, 1, 3])
-        final_shape = op.Shape(transposed, _outputs=["final_shape"], start=0)
-        final_dim = op.Slice(final_shape, [-2], [-1])
+        final_shape = op.Shape(transposed, _outputs=["final_shape"])
+        final_dim = op.Slice(final_shape, start, end)
         return final_dim
 
-    def check(self, context, dim0, dim1, dim2, dim3, final_shape, **_) -> bool:
+    def check(self, context, dim0, dim1, dim2, dim3, final_shape, start, end, **_) -> bool:
```

Comment:
possible to annotate the types?

Reply:
Thought about it. Unfortunately, that would mean this single line turns into probably 12 lines, taking up valuable screen real estate. Eg. , in a laptop, it might mean we don't get to see the checking code below, or the rewrite method below at the same time as the pattern above. All of the relevant parameters above are `ir.Value` ... which is the common/usual case. More interestingly, it is seeing the pattern above that tells us *where* each value comes from (in the original graph pattern that matched). Which is why it more useful to look at the _pattern_ than just the _type_.